File: edukate-storage/src/main/java/io/github/sanyavertolet/edukate/storage/ReadOnlyStorage.java
package io.github.sanyavertolet.edukate.storage;

import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import java.nio.ByteBuffer;

public interface ReadOnlyStorage<Key, Metadata> {
    Mono<Metadata> metadata(Key key);

    Flux<ByteBuffer> getContent(Key key);

    Mono<String> generatePresignedUrl(Key key);

    @SuppressWarnings("unused")
    Flux<Key> prefixed(String rawKeyPrefix);
}

File: edukate-storage/src/main/kotlin/io/github/sanyavertolet/edukate/storage/ReadOnlyStorage.kt
package io.github.sanyavertolet.edukate.storage

import java.nio.ByteBuffer
import reactor.core.publisher.Flux
import reactor.core.publisher.Mono

interface ReadOnlyStorage<Key : Any, Metadata : Any> {
    fun metadata(key: Key): Mono<Metadata>

    fun getContent(key: Key): Flux<ByteBuffer>

    fun getContentWithMetadata(key: Key): Mono<ContentWithMetadata<Metadata>>

    fun generatePresignedUrl(key: Key): Mono<String>

    fun prefixed(rawKeyPrefix: String): Flux<Key>
}
